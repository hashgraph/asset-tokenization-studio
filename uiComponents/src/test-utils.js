import { ChakraProvider, extendTheme } from "@chakra-ui/react";
import { render as testingLibraryRender } from "@testing-library/react";
import React from "react";
import { BasePlatformTheme } from "./Theme/theme";
var DefaultTheme = extendTheme(BasePlatformTheme);
export var render = function (ui, theme) {
    if (theme === void 0) { theme = DefaultTheme; }
    var wrapper = function (_a) {
        var children = _a.children;
        return (React.createElement(ChakraProvider, { theme: theme }, children));
    };
    return testingLibraryRender(ui, { wrapper: wrapper });
};
