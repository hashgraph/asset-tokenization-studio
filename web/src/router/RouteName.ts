export enum RouteName {
  Landing = "LANDING",
  Dashboard = "DASHBOARD",
  AddSecurity = "ADD_SECURITY",
  CreateSecurity = "CREATE_SECURITY",
  CreateEquity = "CREATE_EQUITY",
  CreateBond = "CREATE_BOND",
  DigitalSecurityDetails = "DIGITAL_SECURITY_DETAILS",
  DigitalSecuritiesList = "DIGITAL_SECURITIES_LIST",
  DigitalSecurityMint = "DIGITAL_SECURITY_MINT",
  DigitalSecurityTransfer = "DIGITAL_SECURITY_TRANSFER",
  DigitalSecurityForceTransfer = "DIGITAL_SECURITY_FORCE_TRANSFER",
  DigitalSecurityRedeem = "DIGITAL_SECURITY_REDEEM",
  DigitalSecurityForceRedeem = "DIGITAL_SECURITY_FORCE_REDEEM",
}
