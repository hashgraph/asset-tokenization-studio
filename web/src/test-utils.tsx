import React from "react";
import { ChakraProvider } from "@chakra-ui/react";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import type { RenderOptions, RenderResult } from "@testing-library/react";
import { render, waitFor } from "@testing-library/react";
import type { MemoryHistory } from "history";
import { createMemoryHistory } from "history";
import { I18nextProvider } from "react-i18next";
import { Router } from "react-router-dom";
import i18n from "./i18n";
import theme from "./theme";
import userEvent from "@testing-library/user-event";

export const selectCalendar = async (
  component: RenderResult,
  id: string,
  day: string | number = new Date().getDate(),
) => {
  const calendar = component.getByTestId(id);
  await userEvent.click(calendar);

  await waitFor(() => {
    const daysToSelect = component.getAllByTestId(`day-${day}`);

    const dayToSelect = daysToSelect.find(
      (day) => !day.hasAttribute("disabled"),
    );

    if (dayToSelect) {
      userEvent.click(dayToSelect);
    }
  });

  await userEvent.click(document.body);
};

export const testQueryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
      refetchOnMount: true,
      cacheTime: 0,
    },
  },
  logger: {
    log: console.log,
    warn: console.warn,
    // âœ… no more errors on the console for tests
    error: () => {},
  },
});

export const queryWrapper = ({ children }: { children: React.ReactNode }) => (
  <QueryClientProvider client={testQueryClient}>{children}</QueryClientProvider>
);

const memoryHistory = createMemoryHistory();

export const AllProviders = ({
  children,
  history = memoryHistory,
}: {
  children?: React.ReactNode;
  history?: MemoryHistory;
}) => {
  return (
    <I18nextProvider i18n={i18n}>
      <QueryClientProvider client={testQueryClient}>
        <ChakraProvider theme={theme}>
          <Router location={history.location} navigator={history}>
            {children}
          </Router>
        </ChakraProvider>
      </QueryClientProvider>
    </I18nextProvider>
  );
};

const customRender = (
  ui: React.ReactElement,
  {
    options,
    history,
  }: {
    options?: RenderOptions;
    history?: MemoryHistory;
  } = {},
): RenderResult =>
  render(ui, {
    wrapper: ({ children }) => (
      <AllProviders history={history}>{children}</AllProviders>
    ),
    ...options,
  });

export { customRender as render };
