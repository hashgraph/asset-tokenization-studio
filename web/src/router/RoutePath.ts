import { RouteName } from "./RouteName";

export const RoutePath: Record<RouteName, string> = {
  [RouteName.Dashboard]: "/",
  [RouteName.Landing]: "/connect-to-metamask",
  [RouteName.DigitalSecurityDetails]: "/security/:id",
  [RouteName.DigitalSecurityMint]: "/security/:id/mint",
  [RouteName.DigitalSecurityTransfer]: "/security/:id/transfer",
  [RouteName.DigitalSecurityForceTransfer]: "/security/:id/forceTransfer",
  [RouteName.DigitalSecurityRedeem]: "/security/:id/redeem",
  [RouteName.DigitalSecurityForceRedeem]: "/security/:id/forceRedeem",
  [RouteName.DigitalSecuritiesList]: "/list/:type",
  [RouteName.AddSecurity]: "/security/add",
  [RouteName.CreateSecurity]: "/security/create",
  [RouteName.CreateEquity]: "/security/create/equity",
  [RouteName.CreateBond]: "/security/create/bond",
};
