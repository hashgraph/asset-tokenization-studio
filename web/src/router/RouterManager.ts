import type { ReverseParams } from "named-urls";
import { reverse } from "named-urls";
import { router } from ".";
import { RouteName } from "./RouteName";
import { RoutePath } from "./RoutePath";

export interface RouteParams {
  params?: ReverseParams;
  extra?: string;
  state?: object;
}

export class BaseRouterManager {
  constructor(private routes: Record<RouteName, string> = RoutePath) {}

  #getPath(name: RouteName, { params, extra = "" }: RouteParams) {
    const pattern = this.routes[name] + extra;
    return reverse(pattern, params);
  }

  to(name: RouteName, { state, ...params }: RouteParams = {}) {
    const pathname = this.#getPath(name, params);
    return router.navigate({ pathname }, { state });
  }

  getUrl(name: RouteName, params: RouteParams = {}) {
    const path = this.#getPath(name, params);
    return reverse(path);
  }

  goBack() {
    return router.navigate(-1);
  }

  goDashboard() {
    return this.to(RouteName.Dashboard);
  }
}

export const RouterManager = new BaseRouterManager(RoutePath);
