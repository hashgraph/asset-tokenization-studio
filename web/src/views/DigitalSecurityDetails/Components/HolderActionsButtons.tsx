import { Link as RouterLink, useParams } from "react-router-dom";
import { useTranslation } from "react-i18next";
import { HStack } from "@chakra-ui/react";
import { Button } from "@hashgraph/uiComponents/Interaction";
import { RouteName } from "../../../router/RouteName";
import { RouterManager } from "../../../router/RouterManager";
import { useGetIsPaused } from "../../../hooks/queries/useGetSecurityDetails";
import { PauseRequest } from "@hashgraph/securitytoken-sdk";

export const HolderActionsButtons = () => {
  const { t: tButtons } = useTranslation("security", {
    keyPrefix: "details.actions",
  });
  const { id = "" } = useParams();
  const { data: isPaused } = useGetIsPaused(
    new PauseRequest({ securityId: id }),
  );

  if (isPaused) return <></>;

  return (
    <HStack
      data-testid="holder-actions-buttons"
      w="full"
      justifyContent="flex-end"
      gap={4}
    >
      <Button
        data-testid="transfer-button"
        as={RouterLink}
        to={RouterManager.getUrl(RouteName.DigitalSecurityTransfer, {
          params: { id },
        })}
        variant="secondary"
      >
        {tButtons("transfer")}
      </Button>
      <Button
        data-testid="redeem-button"
        as={RouterLink}
        to={RouterManager.getUrl(RouteName.DigitalSecurityRedeem, {
          params: { id },
        })}
        variant="secondary"
      >
        {tButtons("redeem")}
      </Button>
    </HStack>
  );
};
