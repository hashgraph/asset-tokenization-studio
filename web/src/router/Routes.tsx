import i18n from "i18next";
import { MainLayout } from "../layouts/MainLayout";
import { GenericRoute } from "./components/GenericRoute";
import { PrivateRoute } from "./components/PrivateRoute";
import { RouteName } from "./RouteName";
import { RoutePath } from "./RoutePath";
import { Landing } from "../views/Landing/Landing";
import { AddSecurity } from "../views/AddSecurity/AddSecurity";
import { CreateSecurity } from "../views/Dashboard/Components/CreateSecurity";
import { CreateEquity } from "../views/CreateEquity/CreateEquity";
import { CreateBond } from "../views/CreateBond/CreateBond";
import { Dashboard } from "../views/Dashboard/Dashboard";
import { DigitalSecuritiesList } from "../views/DigitalSecuritiesList/DigitalSecuritiesList";
import { DigitalSecurityDetails } from "../views/DigitalSecurityDetails/DigitalSecurityDetails";
import { DigitalSecurityMint } from "../views/DigitalSecurityMint/DigitalSecurityMint";
import { DigitalSecurityTransfer } from "../views/DigitalSecurityTransfer/DigitalSecurityTransfer";
import { DigitalSecurityRedeem } from "../views/DigitalSecurityRedeem/DigitalSecurityRedeem";
import { DigitalSecurityForceTransfer } from "../views/DigitalSecurityForceTransfer/DigitalSecurityForceTransfer";
import { DigitalSecurityForceRedeem } from "../views/DigitalSecurityForceRedeem/DigitalSecurityForceRedeem";

const t = (key: RouteName) => i18n.t(`routes:${key}`);

export const routes = [
  {
    element: <MainLayout />,
    children: [
      {
        path: RoutePath.LANDING,
        breadcrumb: t(RouteName.Landing),
        element: <Landing />,
      },
      {
        path: RoutePath.DASHBOARD,
        breadcrumb: t(RouteName.Dashboard),
        element: (
          <PrivateRoute>
            <Dashboard />
          </PrivateRoute>
        ),
      },
      {
        path: RoutePath.DIGITAL_SECURITIES_LIST,
        breadcrumb: t(RouteName.DigitalSecuritiesList),
        element: (
          <PrivateRoute>
            <DigitalSecuritiesList />
          </PrivateRoute>
        ),
      },
      {
        path: RoutePath.ADD_SECURITY,
        breadcrumb: t(RouteName.AddSecurity),
        element: (
          <PrivateRoute>
            <AddSecurity />
          </PrivateRoute>
        ),
      },
      {
        path: RoutePath.CREATE_SECURITY,
        breadcrumb: t(RouteName.CreateSecurity),
        element: (
          <PrivateRoute>
            <CreateSecurity />
          </PrivateRoute>
        ),
      },
      {
        path: RoutePath.CREATE_EQUITY,
        breadcrumb: t(RouteName.CreateEquity),
        element: (
          <PrivateRoute>
            <CreateEquity />
          </PrivateRoute>
        ),
      },
      {
        path: RoutePath.CREATE_BOND,
        breadcrumb: t(RouteName.CreateBond),
        element: (
          <PrivateRoute>
            <CreateBond />
          </PrivateRoute>
        ),
      },
      {
        path: RoutePath.DIGITAL_SECURITY_DETAILS,
        breadcrumb: t(RouteName.DigitalSecurityDetails),
        element: (
          <PrivateRoute>
            <DigitalSecurityDetails />
          </PrivateRoute>
        ),
      },
      {
        path: RoutePath.DIGITAL_SECURITY_MINT,
        breadcrumb: t(RouteName.DigitalSecurityMint),
        element: (
          <PrivateRoute>
            <DigitalSecurityMint />
          </PrivateRoute>
        ),
      },
      {
        path: RoutePath.DIGITAL_SECURITY_TRANSFER,
        breadcrumb: t(RouteName.DigitalSecurityTransfer),
        element: (
          <PrivateRoute>
            <DigitalSecurityTransfer />
          </PrivateRoute>
        ),
      },
      {
        path: RoutePath.DIGITAL_SECURITY_FORCE_TRANSFER,
        breadcrumb: t(RouteName.DigitalSecurityForceTransfer),
        element: (
          <PrivateRoute>
            <DigitalSecurityForceTransfer />
          </PrivateRoute>
        ),
      },
      {
        path: RoutePath.DIGITAL_SECURITY_REDEEM,
        breadcrumb: t(RouteName.DigitalSecurityRedeem),
        element: (
          <PrivateRoute>
            <DigitalSecurityRedeem />
          </PrivateRoute>
        ),
      },
      {
        path: RoutePath.DIGITAL_SECURITY_FORCE_REDEEM,
        breadcrumb: t(RouteName.DigitalSecurityForceRedeem),
        element: (
          <PrivateRoute>
            <DigitalSecurityForceRedeem />
          </PrivateRoute>
        ),
      },
    ],
  },
  {
    path: "*",
    element: (
      <PrivateRoute>
        <GenericRoute />
      </PrivateRoute>
    ),
  },
];
