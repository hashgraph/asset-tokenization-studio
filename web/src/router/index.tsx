import { RouterProvider, createBrowserRouter } from "react-router-dom";
import { routes } from "./Routes";
import { useAccountStore } from "../store/accountStore";
import { useEffect } from "react";
import { useSecurityStore } from "../store/securityStore";

export const router = createBrowserRouter(routes);

const AppRouter = () => {
  const { isResetted, reset: resetAccountStore } = useAccountStore();
  const { reset: reserSecurityStore } = useSecurityStore();

  useEffect(() => {
    if (!isResetted) {
      resetAccountStore();
      reserSecurityStore();
    }

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return <RouterProvider router={router} />;
};

export default AppRouter;
