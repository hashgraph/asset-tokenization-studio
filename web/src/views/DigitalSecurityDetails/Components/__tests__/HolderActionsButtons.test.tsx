import { HolderActionsButtons } from "../HolderActionsButtons";
import { render } from "../../../../test-utils";
import userEvent from "@testing-library/user-event";
import { RouteName } from "../../../../router/RouteName";
import { RouterManager } from "../../../../router/RouterManager";

jest.mock("../../../../router/RouterManager", () => ({
  RouterManager: {
    ...jest.requireActual("../../../../router/RouterManager").RouterManager,
    getUrl: jest.fn(),
  },
}));

describe(`${HolderActionsButtons}.name`, () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const factoryComponent = () => render(<HolderActionsButtons />);

  test("should render correctly", () => {
    const component = factoryComponent();

    expect(component.asFragment()).toMatchSnapshot();
  });

  test("should have a transfer button", () => {
    const component = factoryComponent();

    expect(component.getByTestId("transfer-button")).toBeInTheDocument();
  });

  test("transfer button should redirect to transfer page", async () => {
    const component = factoryComponent();

    const button = component.getByTestId("transfer-button");
    await userEvent.click(button);

    expect(RouterManager.getUrl).toHaveBeenCalled();
    expect(RouterManager.getUrl).toHaveBeenCalledWith(
      RouteName.DigitalSecurityTransfer,
      { params: { id: "" } },
    );
  });

  test("should have a redeem button", () => {
    const component = factoryComponent();

    expect(component.getByTestId("redeem-button")).toBeInTheDocument();
  });

  test("redeem button should redirect to redeem page", async () => {
    const component = factoryComponent();

    const button = component.getByTestId("redeem-button");
    await userEvent.click(button);

    expect(RouterManager.getUrl).toHaveBeenCalled();
    expect(RouterManager.getUrl).toHaveBeenCalledWith(
      RouteName.DigitalSecurityRedeem,
      { params: { id: "" } },
    );
  });
});
