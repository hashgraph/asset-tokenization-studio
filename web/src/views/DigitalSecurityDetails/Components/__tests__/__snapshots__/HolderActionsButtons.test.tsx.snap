// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`() => {
  /* istanbul ignore next */
  cov_qodd5a1xd().f[0]++;
  const {
    t: tButtons
  } =
  /* istanbul ignore next */
  (cov_qodd5a1xd().s[12]++, (0, react_i18next_1.useTranslation)("security", {
    keyPrefix: "details.actions"
  }));
  const {
    id =
    /* istanbul ignore next */
    (cov_qodd5a1xd().b[0][0]++, "")
  } =
  /* istanbul ignore next */
  (cov_qodd5a1xd().s[13]++, (0, react_router_dom_1.useParams)());
  const {
    data: isPaused
  } =
  /* istanbul ignore next */
  (cov_qodd5a1xd().s[14]++, (0, useGetSecurityDetails_1.useGetIsPaused)(new asset_tokenization_sdk_1.PauseRequest({
    securityId: id
  })));
  /* istanbul ignore next */
  cov_qodd5a1xd().s[15]++;
  if (isPaused) {
    /* istanbul ignore next */
    cov_qodd5a1xd().b[1][0]++;
    cov_qodd5a1xd().s[16]++;
    return (0, jsx_runtime_1.jsx)(jsx_runtime_1.Fragment, {});
  } else
  /* istanbul ignore next */
  {
    cov_qodd5a1xd().b[1][1]++;
  }
  cov_qodd5a1xd().s[17]++;
  return (0, jsx_runtime_1.jsxs)(react_1.HStack, {
    "data-testid": "holder-actions-buttons",
    w: "full",
    justifyContent: "flex-end",
    gap: 4,
    children: [(0, jsx_runtime_1.jsx)(io_bricks_ui_1.Button, {
      "data-testid": "transfer-button",
      as: react_router_dom_1.Link,
      to: RouterManager_1.RouterManager.getUrl(RouteName_1.RouteName.DigitalSecurityTransfer, {
        params: {
          id
        }
      }),
      variant: "secondary",
      children: tButtons("transfer")
    }), (0, jsx_runtime_1.jsx)(io_bricks_ui_1.Button, {
      "data-testid": "redeem-button",
      as: react_router_dom_1.Link,
      to: RouterManager_1.RouterManager.getUrl(RouteName_1.RouteName.DigitalSecurityRedeem, {
        params: {
          id
        }
      }),
      variant: "secondary",
      children: tButtons("redeem")
    })]
  });
}.name should render correctly 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-1xa9bwv"
    data-testid="holder-actions-buttons"
  >
    <a
      class="chakra-button css-vueqt9"
      data-testid="transfer-button"
      href="/"
    >
      Transfer
    </a>
    <a
      class="chakra-button css-vueqt9"
      data-testid="redeem-button"
      href="/"
    >
      Redeem
    </a>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
